src/persist-client/src/impl/paths.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::ops::Deref;
+use std::str::FromStr;
+use uuid::Uuid;
+
+use crate::r#impl::encoding::parse_id;
+use crate::{ShardId, WriterId};
+
+/// An opaque identifier for an individual blob of a persist durable TVC (aka shard).
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct PartId(pub(crate) [u8; 16]);
+
+impl std::fmt::Display for PartId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "p{}", Uuid::from_bytes(self.0))
+    }
+}
+
+impl std::fmt::Debug for PartId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "PartId({})", Uuid::from_bytes(self.0))
+    }
+}
+
+impl FromStr for PartId {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        parse_id('p', "PartId", s).map(PartId)
+    }
+}
+
+impl PartId {
+    pub(crate) fn new() -> Self {
+        PartId(*Uuid::new_v4().as_bytes())
+    }
+}
+
+/// Partially encoded path used in [mz_persist::location::Blob] storage.
+/// Composed of a [crate::write::WriterId] and [crate::paths::PartId]. Can
+/// be completed with a [crate::ShardId] to form a full [crate::paths::BlobKey].
+///
+/// Used to reduce the bytes needed to refer to a blob key in memory and
+/// in persistent state, all access to blobs are always within the context
+/// of an individual shard.
+#[derive(Clone, Debug, PartialEq)]
+pub struct PartialBlobKey(pub(crate) String);
+
+impl PartialBlobKey {
+    pub fn new(writer_id: &WriterId) -> PartialBlobKey {
```

Comment:
I have a preference to make this deterministic. lets take the PartId in the constructor instead (and just provide `&PartId::new()` in the places it's called) plz